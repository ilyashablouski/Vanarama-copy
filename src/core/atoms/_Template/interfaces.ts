import { IBaseProps } from '../../interfaces/base';

export interface ITemplateProps extends IBaseProps {}
