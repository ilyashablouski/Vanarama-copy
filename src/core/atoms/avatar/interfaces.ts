import { IBaseProps } from '../../interfaces/base';
import { TSize } from '../../../types/size';

export interface IAvatarProps extends IBaseProps {
  src?: string;
  size?: TSize;
  altText?: string;
}
