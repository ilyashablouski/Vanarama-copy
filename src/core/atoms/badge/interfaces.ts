import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IBadgeProps extends IBaseProps {
  size?: TSize;
  color?: TColor;
  label?: string;
}
