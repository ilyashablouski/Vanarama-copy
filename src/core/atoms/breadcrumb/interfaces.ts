import { IBaseProps } from '../../interfaces/base';
import { ILink } from '../../interfaces/link';

export interface IBreadcrumbProps extends IBaseProps {
  items?: ILink[];
}
