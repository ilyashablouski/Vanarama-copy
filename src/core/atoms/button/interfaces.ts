import { ButtonHTMLAttributes } from 'react';
import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IButtonProps
  extends IBaseProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  color?: TColor;
  fill?: 'solid' | 'outline' | 'clear';
  icon?: React.ReactNode;
  iconColor?: TColor;
  iconPosition?: 'before' | 'after';
  label?: string | React.ReactNode;
  round?: boolean;
  size?: TSize;
  withoutDefaultClass?: boolean;
  customCTAColor?: string;
}
