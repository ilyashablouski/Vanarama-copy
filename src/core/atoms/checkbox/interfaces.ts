import { InputHTMLAttributes, ReactElement } from 'react';
import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';

export interface ICheckboxProps
  extends IBaseProps,
    InputHTMLAttributes<HTMLInputElement> {
  /**
   * A color for the checkbox.
   */
  color?: TColor;
  /**
   * A unique ID for the checkbox node - this is required here because the
   * underlying checkbox is hidden and therefore you can only click on the label
   * and the label needs to know which input it corresponds to.
   */
  id: string;
  /**
   * 	Invalid state of the checkbox.
   */
  invalid?: boolean;
  /**
   * A name for the input node.
   */
  label: string | (string | ReactElement)[];
  /**
   * Outlined state of the checkbox.
   */
  outline?: boolean;
  /**
   * A ref to attach to the checkbox.
   */
  ref?: React.Ref<HTMLInputElement>;
}
