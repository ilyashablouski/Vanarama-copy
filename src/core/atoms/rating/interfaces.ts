import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IRatingProps extends IBaseProps {
  size?: TSize;
  color?: TColor;
  score: number;
  max?: number;
  labelColor?: TColor;
  onClick?: (index: number) => void;
  noLabel?: boolean;
}

export interface IIconWrapperProps {
  children: React.ReactNode;
  index: number;
  onClick?: (index: number) => void;
}
