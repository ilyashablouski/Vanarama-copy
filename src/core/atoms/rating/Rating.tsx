import React, { FC, memo } from 'react';
import cx from 'classnames';

import Text from '../text';
import Icon from '../icon';

import StarSharp from '../../assets/icons/StarSharp';
import StarHalfSharp from '../../assets/icons/StarHalfSharp';
import StarOutline from '../../assets/icons/StarOutline';

import { IIconWrapperProps, IRatingProps } from './interfaces';

const IconWrapper: FC<IIconWrapperProps> = memo(
  ({ index, children, onClick }) => {
    return onClick ? (
      <span
        role="button"
        onClick={() => onClick?.(index)}
        style={{ cursor: 'pointer' }}
      >
        {children}
      </span>
    ) : (
      <span>{children}</span>
    );
  },
);

const MIN_SCORE = 0;
const MAX_SCORE = 5;

const Rating: FC<IRatingProps> = memo(props => {
  const {
    className,
    color = 'orange',
    size = 'regular',
    labelColor = 'dark',
    score = 4.5,
    noLabel = false,
    onClick,
    dataUiTestId,
  } = props;

  let i;
  let value;
  const items = [];

  if (score < MIN_SCORE) {
    console.error(`The rating cannot be lower than ${MIN_SCORE}!`);
    value = MIN_SCORE;
  } else if (score > MAX_SCORE) {
    console.error(`The rating cannot be higher than ${MAX_SCORE}!`);
    value = MAX_SCORE;
  } else {
    value = score;
  }

  // Full.
  if (value && value > 0) {
    for (i = 0; i < Math.trunc(value); i += 1) {
      const rate = items.length + 1;
      items.push(
        <IconWrapper key={`full${i}`} index={rate} onClick={onClick}>
          <Icon icon={<StarSharp />} color={color} />
        </IconWrapper>,
      );
    }
  }

  // Half.
  if (value && value > 0 && !Number.isInteger(value)) {
    const rate = items.length + 1;
    items.push(
      <IconWrapper key="half" index={rate} onClick={onClick}>
        <Icon icon={<StarHalfSharp />} color={color} />
      </IconWrapper>,
    );
  }

  // Empty.
  if (value && value > 0 && !Number.isInteger(value)) {
    value += 1;
  }
  for (i = 0; i < Math.ceil(MAX_SCORE - value); i += 1) {
    const rate = items.length + 1;
    items.push(
      <IconWrapper key={`empty${i}`} index={rate} onClick={onClick}>
        <Icon icon={<StarOutline />} color={color} />
      </IconWrapper>,
    );
  }

  return (
    <div
      className={cx('rating', className, `-${size}`, `-${color}`)}
      data-uitestid={dataUiTestId}
    >
      {items}
      {!noLabel && (
        <Text size={size} color={labelColor}>
          {score}
        </Text>
      )}
    </div>
  );
});

Rating.displayName = 'Rating';

export default Rating;
