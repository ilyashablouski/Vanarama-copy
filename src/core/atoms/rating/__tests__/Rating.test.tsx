import React from 'react';
import renderer from 'react-test-renderer';

import Rating from '..';
import { IRatingProps } from '../interfaces';

const optionalProps = {
  id: 'component',
  className: 'class',
  size: 'small' as const,
  color: 'orange' as const,
  max: 6,
};

const mandatoryProps = {
  score: 3.5,
};

function getComponent(props?: Partial<IRatingProps>) {
  return renderer
    .create(<Rating score={mandatoryProps.score} {...props} />)
    .toJSON();
}

describe('<Rating />', () => {
  it('renders correctly', () => {
    const tree = getComponent();
    expect(tree).toMatchSnapshot();
  });

  it('renders without score', () => {
    const tree = getComponent({ ...optionalProps, score: 0 });
    expect(tree).toMatchSnapshot();
  });

  it('should not render without max', () => {
    const tree = getComponent({ ...optionalProps, max: 0 });
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with optional props', () => {
    const tree = getComponent(optionalProps);
    expect(tree).toMatchSnapshot();
  });
});
