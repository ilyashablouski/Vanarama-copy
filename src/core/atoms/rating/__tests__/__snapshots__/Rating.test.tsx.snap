// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rating /> renders correctly 1`] = `
<div
  className="rating -regular -orange"
>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7ZM274.63,347.82,256,334.49V134.39l26,78.91,7.24,22H394.63l-67.32,47.2-19.69,13.81,7.78,22.75,26.26,76.75Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span
    className="text -regular -dark"
  >
    3.5
  </span>
</div>
`;

exports[`<Rating /> renders correctly with optional props 1`] = `
<div
  className="rating class -small -orange"
>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M496,203.3H312.36L256,32,199.64,203.3H16L166.21,308.7,107.71,480,256,373.84,404.29,480,345.68,308.7ZM274.63,347.82,256,334.49V134.39l26,78.91,7.24,22H394.63l-67.32,47.2-19.69,13.81,7.78,22.75,26.26,76.75Z"
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span
    className="text -small -dark"
  >
    3.5
  </span>
</div>
`;

exports[`<Rating /> renders without score 1`] = `
<div
  className="rating class -small -orange"
>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span>
    <i
      className="icon -orange"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M480,208H308L256,48,204,208H32l140,96L118,464,256,364,394,464,340,304Z"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinejoin": "round",
              "strokeWidth": 32,
            }
          }
        />
      </svg>
    </i>
  </span>
  <span
    className="text -small -dark"
  >
    0
  </span>
</div>
`;

exports[`<Rating /> should not render without max 1`] = `null`;
